flag_groups.go

```diff
@@ -0,0 +1,144 @@
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	flag "github.com/spf13/pflag"
+)
+
+const (
+	RequiredAsGroup   = "cobra_annotation_required_if_others_set"
+	MutuallyExclusive = "cobra_annotation_mutually_exclusive"
+)
+
+func (c *Command) MarkFlagsRequiredTogether(flagNames ...string) {
+	for _, v := range flagNames {
+		f := c.Flags().Lookup(v)
+		if f.Annotations == nil {
+			f.Annotations = map[string][]string{}
+		}
+		// Each time this is called is a single new entry; this allows it to be a member of multiple groups if needed.
+		f.Annotations[RequiredAsGroup] = append(f.Annotations[RequiredAsGroup], strings.Join(flagNames, " "))
+	}
+}
+
+func (c *Command) MarkFlagsMutuallyExclusive(flagNames ...string) {
+	for _, v := range flagNames {
+		f := c.Flags().Lookup(v)
+		if f.Annotations == nil {
+			f.Annotations = map[string][]string{}
+		}
+		// Each time this is called is a single new entry; this allows it to be a member of multiple groups if needed.
+		f.Annotations[MutuallyExclusive] = append(f.Annotations[MutuallyExclusive], strings.Join(flagNames, " "))
+	}
+}
+
+// validateFlagGroups validates the mutuallyExclusive/requiredAsGroup logic.
+func (c *Command) validateFlagGroups() error {
+	if c.DisableFlagParsing {
+		return nil
+	}
+
+	flags := c.Flags()
+
+	// groupStatus format is the list of flags as a unique ID,
+	// then a map of each flag name and whether it is set or not.
+	groupStatus := map[string]map[string]bool{}
+	mutuallyExclusiveGroupStatus := map[string]map[string]bool{}
+	flags.VisitAll(func(pflag *flag.Flag) {
+		processFlagForGroupAnnotation(pflag, RequiredAsGroup, groupStatus)
+		processFlagForGroupAnnotation(pflag, MutuallyExclusive, mutuallyExclusiveGroupStatus)
+	})
+
+	errs := validateRequiredFlagGroups(groupStatus)
+	errsExclusive := validateExclusiveFlagGroups(mutuallyExclusiveGroupStatus)
+	errs = append(errs, errsExclusive...)
+
+	if len(errs) > 0 {
+		return combineErrors(errs, `, `)
+	}
+	return nil
+}
+
+func processFlagForGroupAnnotation(pflag *flag.Flag, annotation string, groupStatus map[string]map[string]bool) {
+	groupInfo, found := pflag.Annotations[annotation]
+	if found {
+		for _, group := range groupInfo {
+			if groupStatus[group] == nil {
+				groupStatus[group] = map[string]bool{}
+
+				flagnames := strings.Split(group, " ")
+				for _, name := range flagnames {
+					groupStatus[group][name] = false
+				}
+			}
+
+			groupStatus[group][pflag.Name] = pflag.Changed
+		}
+	}
+}
+
+func validateRequiredFlagGroups(data map[string]map[string]bool) []error {
+	var errs []error
+	for flagList, flagnameAndStatus := range data {
+		unset := []string{}
+		for flagname, isSet := range flagnameAndStatus {
+			if !isSet {
+				unset = append(unset, flagname)
+			}
+		}
+		if len(unset) == len(flagnameAndStatus) || len(unset) == 0 {
+			continue
+		}
+		sort.Strings(unset)
+		errs = append(errs, fmt.Errorf("if any flags in the group [%v] are set they must all be set; missing %v", flagList, unset))
```

Comment:
The working seems ok to me.  I'm just a little unsure about the format when multiple errors are concatenated together.  For example:
```
./prog child --automatic yes --destination a --fast-deploy fish --effective a                                                       <k8s-lab>
Error: if any flags in the group [automatic best certificate] are set they must all be set; missing [best certificate], if any flags in the group [destination fast-deploy] are set none of the others can be; [destination fast-deploy] were all set
Usage:
  prog child [flags]

Flags:
[..]
```
Should the error messages be separated by a newline character?

Reply:
A newline would work for me; I also didnt like the format but tried to follow the general Go guidance of not using capital letters or ending punctuation.

So you'd prefer something like :

```
Error: if any flags in the group [automatic best certificate] are set they must all be set; missing [best certificate]
if any flags in the group [destination fast-deploy] are set none of the others can be; [destination fast-deploy] were all set

Usage:...
```

I could also try and make it so that it makes it like a bulleted list? They are long errors so it doesnt flow particularly great if you have multiples:

```
Errors: 
 - if any flags in the group [automatic best certificate] are set they must all be set; missing [best certificate]
 - if any flags in the group [destination fast-deploy] are set none of the others can be; [destination fast-deploy] were all set

Usage:...

```

I think the latter looks the best for and end user on the CLI and is just a tiny bit more verbose to implement. What do you think?

Reply:
Why do you have to return all errors? The normal flag parsing code only returns the first error it encounters. This should simplify the code and make the errors smaller and thus better readable.

Reply:
> Why do you have to return all errors? The normal flag parsing code only returns the first error it encounters. This should simplify the code and make the errors smaller and thus better readable.

I guess it was to avoid a user having to fix one error at a time, if there were multiple errors?
I see value in both approaches personally, so I'll defer to what you guys settle on.

That being said, the current code is missing the first newline when there are multiple errors:
```
Error: - if any flags in the group [automatic best certificate] are set they must all be set; missing [best certificate]
- if any flags in the group [destination fast-deploy] are set none of the others can be; [destination fast-deploy] were all set
```

Reply:
I'm definitely fine with the simplification. Not sure why I defaulted to that originally, I sometimes use both. Making the simplification avoids that problem all together and since its just at invocation time and feedback is fast I dont think its a problem if they have to tweak it a few times and keep finding 'new' problems.